### ISSUE TITLE
Firefox support

### ISSUE THREAD

**Issue by un71** [privacy-tech-lab/optmeowt-browser-extension#117]
I am just opening this issue so people can watch it and be notified when it is on AMO.

**Comment by SebastianZimmeck** [comment]
@asalia-1 and @kalicki1, could one of you look into the following by Tuesday? 

I am not able to get OptMeowt v1.1.1 validated as a Firefox Add-on. There seems to be some issue with the yaml library we are using. Maybe, we need to bump the library, use a different one, or (if there is no other way) open a PR on their site to fix any issues that prevent validation.

Could you create a Firefox developer account yourself, try to validate OptMeowt, and see whether you can fix the syntax errors? I do not think you need to submit OptMeowt actually. You can just run the validation. I have not found a way to add multiple users to the privacy-tech-lab developer account.

<img width="1062" alt="Screen Shot 2020-11-22 at 7 07 44 PM" src="https://user-images.githubusercontent.com/11034003/99921099-24729c80-2cf6-11eb-9997-fa901717b513.png">

**Comment by SebastianZimmeck** [comment]
Also, on a side note, for every release:

1. Let's update the version number in the manifest
2. Upload the release on both the Chrome and Firefox stores

**Comment by asalia-1** [comment]
I will look into this

**Comment by kalicki1** [comment]
@asalia-1 and I took a look at this, and we decided that maintaining support for our YAML files will take more work that it's worth for now. It's going to be easier for us to fall back to JSON mainly because its supported so well by default. I will push these changes here in a new branch, and we can merge with the main once we are sure everything works. Then, when we are ready we can update our releases and push to AMO and the Chrome store.

**Comment by kalicki1** [comment]
These should be the changes necessary to pass the issues you had above @SebastianZimmeck. I believe the warnings will still be there, but I think we can go ahead and try to upload the extension and work on fixing them in the meantime.

**Comment by SebastianZimmeck** [comment]
Thanks, @kalicki1! I agree. We can address the warnings later.

Could you and/or @asalia-1 run some tests to make sure OptMeowt is still working (in particular, with regards to the domain list; e.g., selecting individual sites to not receive a GPC signal and exporting/importing the domain list).

Then, bump up the version number in the manifest and update the v1.1.1 release with new packaged files (I guess we could also create a new release, but maybe that looks a little odd on the Chrome Web Store if there is one release missing).

**Comment by kalicki1** [comment]
I just ran a few tests and it looks like everything is working well! I'll open a pull request on this issue, and once it's all set I'll update the v1.1.1 release.

**Comment by kalicki1** [comment]
@SebastianZimmeck, I updated the versioned release v1.1.1. I think we should be all set to continue the process of uploading to AMO.

**Comment by SebastianZimmeck** [comment]
OptMeowt v1.1.1 is under review in the Firefox Add-on and Chrome Web Store. I will update this issue once the reviews are finished.

**Comment by asalia-1** [comment]
During a manual review for OptMeowt v1.1.1 a bug was encountered. This is a snippet of the error logs produced by the bug. 

grunt
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module &#39;js-yaml&#39;
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:690:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.&lt;anonymous&gt; (/home/test/Downloads/optmeowt-1.1.1-src/optmeowt-browser-extension-941d5b8/node_modules/grunt/lib/grunt/file.js:25:12)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)

This is a case of the ys-jaml library not being found and I believe this occured because we removed js-yaml library from the package-lock.json file which is a grunt dependency. I implemented a fix for this issue in this pr #122 . @kalicki1 , I requested your review for this fix. 
#

**Comment by SebastianZimmeck** [comment]
I created a new [release v1.1.2](https://github.com/privacy-tech-lab/optmeowt-browser-extension/releases/tag/v1.1.2) with the code as it currently stands and submitted it to the Firefox Add-on store. Fingers crossed!

I will also update the version on the Chrome Web Store. However, as v1.1.1 is currently still under review I am not able to submit v1.1.2 right now.

**Comment by SebastianZimmeck** [comment]
Submitted v1.1.2 on the Chrome Web Store.

**Comment by SebastianZimmeck** [comment]
Approved on the Chrome Web Store but still pending on AMO.

**Comment by SebastianZimmeck** [comment]
There are various issues that were raised on AMO that @kalicki1 and @asalia-1 will take two weeks or so to fix. I can address point 4 on the privacy policy.

> 1) Extensions defining a content security policy that allows eval (&#39;unsafe-eval&#39;) are generally not allowed for security and performance reasons. eval is only necessary in rare cases. Please use a different method or explain why eval is required in your add-on.
>
> manifest.json - line 37
>
> 2) We don&#39;t allow add-ons to use remote scripts because they can create serious security vulnerabilities. We also need to review all add-on code, and this makes it much more difficult. Please insert those scripts locally from your add-on code.
>
> manifest.json - line 37
options\options.html - lines 29,30
>
> 3) Your add-on includes a third-party library. Please provide the origin of the exact library version you were using and make sure you are using an exact copy of the original maintainers release version. For more information, refer to https://extensionworkshop.com/documentation/publish/third-party-library-usage/ .
>
> libs-js folder
>
> 4) Please add a privacy policy to this add-on that details which user data is being sent and to what services. The critical things to describe in the policy are how your extension collects, uses, stores, and shares or discloses information about people.If your add-on makes it apparent to websites that it is installed, this must also be mentioned. The policy should be about the extension only, not a copy of the website&#39;s privacy policy. It should also be the actual text, as opposed to a link to a privacy policy on a website. The privacy policy can be added in the add-on settings under “Manage Authors &amp; License” on AMO.

**Comment by kalicki1** [comment]
I have been looking into the issues AMO wrote about to try and get our extension back on the Mozilla store. @SebastianZimmeck, since you mentioned you could look into the privacy policy issue in (4), I decided to skip that for the time being. I went through our code, restructured our dependencies to fully rely on the npm system we set up earlier last semester (aside from Dark Mode Switch, which has its own quirks). Let me elaborate what I did to address the first three points:

> 1. Extensions defining a content security policy that allows eval ('unsafe-eval') are generally not allowed for security and performance reasons.
> 2. We don't allow add-ons to use remote scripts because they can create serious security vulnerabilities. We also need to review all add-on code, and this makes it much more difficult. Please insert those scripts locally from your add-on code.

Both of these overlap, so I'll address them together. I decided to cut out the `unsafe-eval` content security policy altogether, since it would have been difficult to justify and work around this permission. In the same vein, I decided to remove the entire content security policy permissions altogether, though to make this work I also had to remove all of our calls to remote scripts. This involved removing CDN calls, moving everything over to our locally managed npm system combined with Grunt. I had to do some testing to make sure `popper.js` and `tippy.js` worked, but I got everything straightened out and these first two issues should be all set. 

> 3. Your add-on includes a third-party library. Please provide the origin of the exact library version you were using and make sure you are using an exact copy of the original maintainers release version.

Moving everything over to npm should solve the "exact copy of the original maintainers release version" issue AMO mentioned in our notes. Dark Mode Switch is still an exception though because I had to modify the code in order to get it to work with our extension. Fortunately, the code is short enough that I hope the reviewers will be able to hand check it if we leave a note. The link in the AMO review above mentioned we should leave the versions we use in a note under `Notes to Reviewers` in the AMO submission page. Here are my notes you can use for the resubmission @SebastianZimmeck.  

```For AMO Notes to Reviewers

[animate.css v4.1.1](https://github.com/animate-css/animate.css/tree/v4.1.1)
[FileSaver.js v2.0.4](https://github.com/eligrey/FileSaver.js/tree/v2.0.4)
[mustache.js v4.1.0](https://github.com/janl/mustache.js/tree/v4.1.0)
[popperjs/core v2.6.0](https://github.com/popperjs/popper-core/tree/v2.6.0)
[psl - Public Suffix List v1.8.0](https://github.com/lupomontero/psl/tree/v1.8.0)
[tippy.js v2.6.7](https://github.com/atomiks/tippyjs/tree/v6.2.7)
[uikit v3.6.9](https://github.com/uikit/uikit/tree/v3.6.9)
[Dark Mode Switch v1.0.0](https://github.com/coliff/dark-mode-switch/tree/v1.0.0)

Note: Our locally saved copy of Dark Mode Switch in the `libs` folder (not `libs-js` which is from NPM) is modified to work with our extension. Since the source code is so short, we figured this is okay for review purposes. 
```

I am also not sure if we had already done this, but the guides I used mentioned we should provide extension build instructions for the reviewers. Just in case, we should include our readme as "extension build instructions" if we haven't done so already. 

Here are some AMO guidelines I used for reference: 
[AMO Extension Workshop - 3rd Party Library Usage](https://extensionworkshop.com/documentation/publish/third-party-library-usage/)
[AMO Extension Workshop - Source Code Submission](https://extensionworkshop.com/documentation/publish/source-code-submission/)

**Comment by SebastianZimmeck** [comment]
> @SebastianZimmeck, since you mentioned you could look into the privacy policy issue in (4), I decided to skip that for the time being.

Here is a privacy policy. It is very short because we do not collect any data. 

```
Privacy Policy
Last updated: January 11, 2021

OptMeowt does not collect any personal information from you. OptMeowt works by sending websites you visit Global Privacy Control (GPC) signals to opt your browser out from the sharing and selling of personal information. This signal is standardized and not in any way personalized to you or your browser. In OptMeowt's settings you can select which sites should receive a GPC signal. OptMeowt also places opt out cookies in your browser as some sites require those to opt you out. 

Please contact us at sebastian@privacytechlab.org for further information
```

You can post this under https://www.privacytechlab.org/optmeowt/privacy.html and ideally make the email address such that it will not be spammed.

**Comment by kalicki1** [comment]
We got approved! 🎉

Since we are on AMO now, we should be all set with this issue. If we have any other specific problems in the future, we can open a new issue to address it.

**Comment by andreicristianpetcu** [comment]
This is sooo coool. I already started using it on Android. Thank you!
![Screenshot_20210114-003219_Firefox_Nightly](https://user-images.githubusercontent.com/922368/104518318-3c80e100-5600-11eb-8f94-4784daff55dc.png)
![Screenshot_20210114-003313_Firefox_Nightly](https://user-images.githubusercontent.com/922368/104518320-3d197780-5600-11eb-9668-ebffb5477373.png)
![Screenshot_20210114-003327_Firefox_Nightly](https://user-images.githubusercontent.com/922368/104518321-3db20e00-5600-11eb-8d81-71f76f9539f4.png)
